kvserver: stop using the cached span config

andrewbaptist opened the pull request:
This change removes almost all uses of the cached span config from the replica. The cached span config is removed in a future commit.

Epic: none

Release note: None

Commit:
kvserver: stop using the cached span config

This change removes almost all uses of the cached span config from the
replica. The cached span config is removed in a future commit.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116796)
<!-- Reviewable:end -->

Commit:
kvserver: remove FallbackSpanConfig

The FallbackSpanConfig no longer makes sense now that we have the
SpanConfig architecture creating a span for every possible key. This
path would only happen on errors and we should never set the
FallbackSpanConfig in practice.

Epic: none

Release note: None

Commit:
kvserver: remove direct use of repl.mu.conf

The repl.mu.conf will be removed and tests that use it need to change to
set the span config differently.
This commit removes it in TestChooseRangeToRebalanceAcrossHeterogeneousZones.

Epic: none

Release note: None

Commit:
kvserver: simplify TestAllocatorCheckRangeUnconfigured

The test was previously using a mock SpanConfigReader to return the
necessary SpanConfig for the different uses. Simplify the test to set
the DefaultSpanConfig instead.

Epic: none

Release note: None

Commit:
kvserver: add utility method to get the SpanConfig

A number of different places in the code need the SpanConfig and there
are subtle differences in the conditions they check when loading it.
The Store.GetSpanConfigForKey method will be the exclusive way that all
places load the SpanConfig.

Epic: none

Release note: None

Commit:
kvserver: pass SpanConfig to split and scatter

The code previously reloaded the descriptor and the SpanConfig between
deciding to split and executing the split. This change changes
additional methods to use the same SpanConfig for the decision for an
action as well as the action itself.

Epic: none

Release note: None
